Fix bucket keys format for range aggregations on float field

salvatore-campagna opened the pull request:
Keys derived from float values seem incorrect due to some rounding
issue and they do not (always) match the original ranges as specified
in the aggregation query. Ideally the user expects keys in the response
to match ranges in the request. When using double values, keys match
expected values.

The PR includes a possible fix. The idea is to generate a key without
using the stored float values. Using stored float values results in
double/float precision issues.

This fix address the bug reported in #81749.

salvatore-campagna committed:
Test: add tests for float and double range aggregation keys

Keys derived from float values seem incorrect due to some rounding
issue and they do not always match the original ranges as specified
in the aggregation query. Ideally the user expects keys in the response
to match ranges in the request. When using double values, keys match
expected values.

These testsq covers a bug reported in #81749.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

salvatore-campagna committed:
Fix bucket key generation for range queries using floats

The idea is to generate the key using the `from` and `to` values
from double values instead of using the stored float values.
The stored float values can still be used for all bucket comparisons.

This fix addresees issue #81749.

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/RangeAggregationBuilder.java:
```diff
@@ -171,7 +171,8 @@ protected RangeAggregatorFactory innerBuild(
                 to = parser.parseDouble(range.toAsStr, false, context::nowInMillis);
             }
-            return new Range(range.key, from, range.fromAsStr, to, range.toAsStr);
+            String key = generateKey(range.from, range.to, config.format());
```
This needs to be done for other aggregation builders too.

salvatore-campagna committed:
Fix key generation for other range aggregators

This includes two changes:
* Use the range key if one is specified for the range
* Use the key generation logic in the date histogram aggregator

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java:
```diff
@@ -52,7 +52,7 @@ public Bucket(
         ) {
             this.keyed = keyed;
             this.format = format;
-            this.key = key != null ? key : generateKey(from, to, format);
```
Not assigning the key here breaks the wire format with a null key 👎

salvatore-campagna committed:
Make sure we always generate a key

Setting a null key results in wire format serialization NPE.
Here we make sure we always have a non-null key.

salvatore-campagna committed:
Fix code format violations

salvatore-campagna committed:
Remove semicolon on new line

nik9000 reviewed (commented):
The fix makes sense to me, but I think it's worth changing the serialization logic so range is no longer optional. the BWC handling code will make it a little more obvious what the upgrade behavior is.

flash1293 commented:
I have no idea how Elasticsearch works so apologies if this is a stupid question, but is this also fixing the `from` and `to` properties in the bucket? These are the ones important to the Kibana use case, we don't check the string bucket name
```
"6.0-10.600000381469727": {
            "from": 6,
            "to": 10.600000381469727,
            "doc_count": 2
          }
```

This is the part causing the issue:
```
            "to": 10.600000381469727,
```

It seems like the code is not addressing this

nik9000 commented:
It isn't changing those, no. Maybe we should rework this whole thing then.
Instead of sending the bounds maybe we need to use the displayed bounds and
real bounds or something.

On Fri, Dec 17, 2021, 12:30 PM Joe Reuter ***@***.***> wrote:

> I have no idea how Elasticsearch works so apologies if this is a stupid
> question, but is this also fixing the from and to properties in the
> bucket? These are the ones important to the Kibana use case, we don't check
> the string bucket name
>
> "6.0-10.600000381469727": {
>             "from": 6,
>             "to": 10.600000381469727,
>             "doc_count": 2
>           }
>
> This is the part causing the issue:
>
>             "to": 10.600000381469727,
>
> It seems like the code is not addressing this
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/81801#issuecomment-996896649>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUXIVQZZNNF5DMOZ733VTURNXUXANCNFSM5KF35JOQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

salvatore-campagna commented:
No it is not actually changing those values in both `to` and `from`.

imotov commented:
Yeah, I am afraid fixing bucket name without fixing the ranges might produce more confusion that it would resolve. I tend to agree with @nik9000 here, we should rethink it here.

salvatore-campagna committed:
fix: use `originalFrom` and `originalTo` fields to keep the original values for range boundaries

When dealing with float fields we actually store float values with lower precision.
Later on, when loading float values into double values the precision difference between
float and double values surfaces as additional "spourious" digits in the string
representation which is propagated to the client during serialization. As a result of this,
the JSON response returned to the client includes range values not matching values in the
request. This results in clients, including Kibana, to break while trying to match ranges in
the request with ranges in the response.

With this change we use two new fields, `originalFrom` and `originalTo` to hold the original
double values before manipulating the precision of float fields. Later we use these fields in
the code dealing with serialization. The effect is that "precision artifacts" resulting from
float to double conversions are not propagated to clients.

salvatore-campagna committed:
fix: use larger precision when comparing double values

salvatore-campagna commented:
@elasticmachine elasticsearch-ci/bwc

salvatore-campagna commented:
@elasticsearchmachine update branch

salvatore-campagna committed:
Merge branch 'master' into test/add-range-agg-key-test

salvatore-campagna commented:
Made a PR which should fix the issue. I see some "Mixed cluster" tests failing. Will try to fix this later today.

salvatore-campagna committed:
Fix mixed cluster tests

salvatore-campagna committed:
fix: skip versions before 8.1.0

salvatore-campagna committed:
fix: update skip version reason

salvatore-campagna committed:
fix: write optiona double values

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/bwc

salvatore-campagna commented:
@elasticmachine  run elasticsearch-ci/part-1

salvatore-campagna committed:
fix: use Double instead of double

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/bwc

imotov commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java:
```diff
@@ -157,7 +172,9 @@ private static String generateKey(double from, double to, DocValueFormat format)
         public void writeTo(StreamOutput out) throws IOException {
             out.writeString(key);
             out.writeDouble(from);
+            out.writeOptionalDouble(originalFrom);
```
You need to add version check on the sending side as well. Otherwise, you will be sending something that another (older) side is not expecting to receive.

salvatore-campagna replied:
I thought it was not necessary because:
 - nodes running version 8.1.0 would write the additional fields
 - nodes not running version 8.1.0 would not write the additional fields

salvatore-campagna committed:
fix: write `originalFrom` and `originalTo` only if version is 8.1.0

salvatore-campagna committed:
test: add test for date range histogram keys

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java:
```diff
@@ -36,25 +37,31 @@
             double from,
+            Double originalFrom,
             double to,
+            Double originalTo,
```
Could this one be always specified - a `double` instead of `Double`?

salvatore-campagna replied:
Will see

salvatore-campagna committed:
fix: use double instead of Double for originalFrom and originalTo

salvatore-campagna committed:
fix: use optional doubles for the wire protocol serialization

salvatore-campagna committed:
Revert "fix: use optional doubles for the wire protocol serialization"

This reverts commit 4ef1da69fd840dfa0be9dc44d01f8b7e9fd598cf.

salvatore-campagna committed:
Revert "fix: use double instead of Double for originalFrom and originalTo"

This reverts commit b29fddd54362bf2a302d83cd0b3202247366a38d.

salvatore-campagna committed:
fix: use double instead of Double for originalFrom and originalTo

not-napoleon commented:
If you're intending to backport this, the custom is to put the all the versions you'll backport as labels on this PR.  That makes it easier later on to see which versions something is in.  From the discussion on the original issue, it seems like this should land in 8.0 and 7.17 as well?

not-napoleon reviewed (approved):
This looks good to me.  I'd appreciate some javadoc on the new fields to indicate why we're doing this, because I don't think it'll be obvious a year from now.  Thanks for fixing it!

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into test/add-range-agg-key-test

salvatore-campagna commented:
I will add the JavaDocs.

salvatore-campagna committed:
docs: improve java docs clarifying usage of originalFrom and originalTo

salvatore-campagna committed:
docs: rephrase java docs

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into test/add-range-agg-key-test

salvatore-campagna committed:
fix: linee longer than 140 characters

salvatore-campagna commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/82434">](https://github.com/elastic/elasticsearch/pull/82434)|
|✅|7.17|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/82435">](https://github.com/elastic/elasticsearch/pull/82435)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
